lib/oli/delivery/attempts/core.ex

```diff
@@ -259,6 +260,57 @@ defmodule Oli.Delivery.Attempts.Core do
     )
   end
 
+  @doc """
+    Retrieves all graded resource access where the last lms grade sync failed
+    for a given section.
+
+    `[
+      %{
+        id: 21390,
+        resource_id: 1234,
+        user_id: 558,
+        user_name: "Some user name",
+        page_title: "Some page title"
+      },
+      ...
+    ]`
+  """
+  def get_failed_grade_sync_resource_accesses_for_section(section_slug) do
+    Repo.all(
+      from(
+        resource_access in ResourceAccess,
+        join: section in assoc(resource_access, :section),
+        join: user in assoc(resource_access, :user),
+        join: section_project_publication in assoc(section, :section_project_publications),
+        join: published_resource in PublishedResource,
+        on: section_project_publication.publication_id == published_resource.publication_id,
+        join: revision in Revision,
+        on: revision.id == published_resource.revision_id and revision.resource_id == resource_access.resource_id,
+        where:
+          section.slug == ^section_slug and
+          section.status == :active and
+          revision.deleted == false and
+          revision.graded == true and
+          (not is_nil(resource_access.last_grade_update_id) and
+            resource_access.last_grade_update_id != resource_access.last_successful_grade_update_id),
```

Comment:
Could checking `last_grade_update_id` against `last_successful_grade_update_id` make it so that it does not retrieve a grade update failure when it's the fist one on the resource?

Tried making the test in `test/oli/delivery/attempts_test.exs:465` only insert a failing update, without a previous successful one, and `Attempts.get_failed_grade_sync_resource_accesses_for_section/1` does not return that failure.

Does that make sense, or do we always have a successful grade update to begin with?

Reply:
You are right, not taking into account that scenario. Changing it...

Thanksss !